### PR TITLE
Implements round-randomized space map

### PR DESCRIPTION
**Pull request by Razharas**
Implements round-randomized torus space map
Kills the space cube
Adds config file for maps, so you can add more with just some non-code
edit

Torus can be very unusual
Like H shape for example
All the "legs" of such torus will loop onto themselves in x axis but will have a big 3 zleves long loop in y axis
The "-" of such torus will have a self-loop in yaxis but 3 zlevel long loop in x axis
While we dont have enough zlevels for exactly H shape we can have any conbination of zlevels that touch at least one other zlevel with their edge
So 6 zveles long loop that looks like a line in x axis is a possibility, or a brick 2x3, or L shape, or L shape with last zlevel of long side not straight but to the side

example
For example this is one of the possible space maps
Red lines represent one zlevel loops(zlevels looping onto themselves)
Blue lines represent 2 zlevel long loops, means you cycle through 2 zlevels in one direction
Green line represent 3 zlevel loops, means you fly through 3 zlevels before returning to the starting one
Max loop length = max linked zlevels(6 in our case)
The picture is of tripple torus, do not be decieved that it is in not a torus since it doesnt look like a doughnut

torus extended

Heres the explanation why is it even called a torus in the first place
The picture is the mapping of the same random space grid i used as example onto the 3d figure, shown from 2 perspectives, as you can see all the lines(2 green, 1 blue and 3 red) which were just lines on the grid are now circles, this explains the looping that happens if you go along these lines
Keeping such unusual loopings while also keeping the general direction changes(like the lines on the grid intersecting at right angle and only once) is only possible on the 3d figure of torus, hence the "torus this" and "torus that"
Also as you can see only way to fit those real strange loopings is to add more holes to that 3d torus
I hope its now clear that its simple and not some magic

### CODE REVIEW

**Comment by duncathan** [comment]
ASSERT(cube == CUCKED)

**Comment by Miauw62** [comment]
what is this make a better fucking op raz

**Comment by Miauw62** [comment]
if you want a torus we need to add 3 more z-levels really

File: code/modules/space transition/space_transition.dm [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+	var/name = "Your config settings failed, you need to fix this for the datum space levels to work"
+	var/list/neigbours
+	var/z_value = 1 //actual z placement
+	var/linked = 1
+	var/xi
+	var/yi   //imaginary placements on the grid
+
+/datum/space_level/New()
+	neigbours = list()
+	var/list/L = list("1","2","4","8")
```
Comment (duncathan) [inline code/modules/space transition/space_transition.dm]: powerful magic you're working here

Comment (duncathan) [inline code/modules/space transition/space_transition.dm]: (i'm pretty sure you could use some defines here)

**Comment by Razharas** [comment]
You can make torus of unusual shape that while retains the looping is not necessery straight and can loops in different places

File: code/modules/space transition/space_transition.dm (hunk 1 above)
Comment (Razharas) [inline code/modules/space transition/space_transition.dm]: Yes probably

**Comment by Miauw62** [comment]
yes i know but i mean from a gameplay standpoint, else space exploration may become a little /too/ easy.

**Comment by Razharas** [comment]
Torus can be very unusual
Like H shape for example
All the "legs" of such torus will loop onto themselves in x axis but will have a big 3 zleves long loop in y axis
The "-" of such torus will have a self-loop in yaxis but 3 zlevel long loop in x axis
While we dont have enough zlevels for exactly H shape we can have any conbination of zlevels that touch at least one other zlevel with their edge
So 6 zveles long loop that looks like a line in x axis is a possibility, or a brick 2x3, or L shape, or L shape with last zlevel of long side not straight but to the side

**Comment by Razharas** [comment]
example
For example this is one of the possible space maps
Red lines represent one zlevel loops(zlevels looping onto themselves)
Blue lines represent 2 zlevel long loops, means you cycle through 2 zlevels in one direction
Green line represent 3 zlevel loops, means you fly through 3 zlevels before returning to the starting one
Max loop length = max linked zlevels(6 in our case)

**Comment by Miauw62** [comment]
neat

**Comment by Razharas** [comment]
Updoot the op post with the information i commented

**Comment by Dorsisdwarf** [comment]
I don't approve of all those infinite loops, frankly. That was the whole complaint of spacecube.

**Comment by WJohn** [comment]
I have no idea what this does to help deal with the spacecube nonsense. It doesn't seem any more or less intuitive since it's all randomly placed anyway.

**Comment by Razharas** [comment]
On one hand its much more random
On the other you are guarandeed to hit the place you left if you travel in straight direction

**Comment by xxalpha** [comment]
death to looping space cube

**Comment by Incoming5643** [comment]
:man: :gun: :cop: 
Don't you fucking dare touch my monstrous code baby that I refuse to pay child support for.

Just kidding, we cool. But I can't really agree with this idea when there's one z-level loops, we should add enough z-levels to make sure that a world torus can be made that doesn't have them (even if that just means more two z-level loops)

**Comment by Razharas** [comment]
Complete lack of one zlevel loops 2will make space navigation too easy

**Comment by Incoming5643** [comment]
How do you figure? If you're drifting through space on a z level with a y-axis loop it just means that you'll only have three connections to plot for the z level instead of four. Strictly speaking that means it's easier to navigate once you've established there's a loop. Any space cartographer worth his space salt will have a GPS to check his z level, so self loops are pretty obvious to figure out if you've prepared.

Also one zlevel loops is a pretty rough nerf to spacing bodies from the station if it happens to be placed somewhere with one.

File: code/modules/space transition/space_transition.dm (hunk 1 above)
Comment (RemieRichards) [inline code/modules/space transition/space_transition.dm]: yeah you never actually use the directions in their number form in this code (as far as I can see, it's late) so you can just swap them for defines of strings (the default direction defines are numbers, and will make byond cry by thinking they're indexes)

eg:
```
#define Z_NEIGHBOUR_NORTH "north"
#define Z_NEIGHBOUR_EAST "east"
```
etc.

**Comment by optimumtact** [comment]
so blow your corpse up

**Comment by JJRcop** [comment]
Could someone expand the amount of z-levels infinitely or is there an eventual cap? With the space cube there were 6 and that was it, it's a cube, but with this, is it infinitely scalable?

**Comment by Incoming5643** [comment]
I don't think there's any easily attainable limit to the number of empty space z-levels, since they're designed to be as low impact as possible in terms of upkeep (so long as starlight isn't on). Only thing they would make worse is turf in world checks, which aren't exactly something to use to begin with (I remember someone running the numbers that showed that for turf checks there was some proc much more efficient than world, but now I forget what it was [whoops]).

**Comment by Razharas** [comment]
System in this pr only need you to add additional zlevel info to the txt file in a special manner and thats it
Apart from this you can add any amount of connected zlevels and the system will adjust intself and build a bigger space map on a bigger grid

**Comment by Cheridan** [comment]
It's not hard to add more space z-levels if it's needed/desired

we could even add

space
content

:scream:

**Comment by optimumtact** [comment]
>inb4 someone uses this as an excuse to add more empty z levels

i'll kill them

**Comment by Iamgoofball** [comment]
>space content

i'd love to but people will meme the fuck out over it

**Comment by Jordie0608** [comment]
What happens if the number of z levels on a map is changed but space_levels.txt isnt? A single config file for multiple maps is the sort of thing I can envision causing problems, probably more for creative mappers than anything else, although I do doubt it'd come up. Could it be possible to associate those configs with the map .dms?

**Comment by Razharas** [comment]
It will runtime durning transition assigmet and zlevel not defined in the txt will be like current centcomm
Centcomm after this pr will loop onto itself in all directions

**Comment by Razharas** [comment]
torus extended

Heres the explanation why is it even called a torus in the first place
The picture is the mapping of the same random space grid i used as example onto the 3d figure, shown from 2 perspectives, as you can see all the lines(2 green, 1 blue and 3 red) which were just lines on the grid are now circles, this explains the looping that happens if you go along these lines
Keeping such unusual loopings while also keeping the general direction changes(like the lines on the grid intersecting at right angle and only once) is only possible on the 3d figure of torus, hence the "torus this" and "torus that"
Also as you can see only way to fit those real strange loopings is to add more holes to that 3d torus
I hope its now clear that its simple and not some magic

**Comment by Miauw62** [comment]
the only real requirements i have for space transitions is that they are reproducable within the same round but very random between rounds.

**Comment by Razharas** [comment]
Thats exactly what this system does

**Comment by Incoming5643** [comment]
To be fair though space cube met those requirements too. People just got flustered because it wasn't exactly compatible with euclidean space.

**Comment by Miauw62** [comment]
i forgot to mention that going back in the direction you came should put you at the same place you started out at.

**Comment by Razharas** [comment]
Technically its not guaranteed if you use some special angles but since our zlevels are 256x256 instead of infiniteXinfinite amount of infinitesmalls it is guaranteed to put you back on the same turf in this or less than (256^2) * 6(amount of linked zlevels) moves

**Comment by JJRcop** [comment]
@Incoming5643 I thought the space cube was actually kinda broken in that it wasn't even functioning as it was supposed to/intended to.

**Comment by Incoming5643** [comment]
It never broke, it was just never perfect to begin with

**Comment by Razharas** [comment]
@Cheridan 
@Aranclanos 
@Miauw62 
@RemieRichards 
@Jordie0608 
BUY YOUR POWERS COMBINED I ASK THIS PR TO BE MERGED
Or at least reviewed

**Comment by Jordie0608** [comment]
Fix the stuff I mentioned dork.

**Comment by Razharas** [comment]
I can try but it will need a strict grammar rules, any unusual bullshit in the dm and it will tumble down

**Comment by Razharas** [comment]
After some attempts i cant imagine a way to do this in a non-retarded way(like comments deciding if map is linkable or not)
So separate txt is the best way about this

**Comment by phil235** [comment]
Tested it, seems to work as advertised. @Razharas Can you add a changelog maybe?

**Comment by Razharas** [comment]
@phil235 
Changelog up
MERGE

**Comment by KorPhaeron** [comment]
This is still horribly broken

**Comment by optimumtact** [comment]
you mean working well

**Comment by freerealestate** [comment]
oh great, back to looping through the same z-levels endlessly without realising because your perspective is so small compared to the size of the z-levels

**Pull request merged**